core/headerchain.go

```diff
@@ -591,353 +486,124 @@ func (hc *HeaderChain) CurrentHeader() *types.Header {
-	// before head header is updated. The method will return the actual block it
-	// updated the head to (missing state) and a flag if setHead should continue
-	// rewinding till that forcefully (exceeded ancient limits)
-	UpdateHeadBlocksCallback func(ethdb.KeyValueWriter, *types.Header) (uint64, bool)
-
-	// DeleteBlockContentCallback is a callback function that is called by SetHead
-	// before each header is deleted.
-	DeleteBlockContentCallback func(ethdb.KeyValueWriter, common.Hash, uint64)
-)
+// Config retrieves the header chain's chain configuration.
+func (hc *HeaderChain) Config() *params.ChainConfig { return hc.config }
 
-// SetHead rewinds the local chain to a new head. Everything above the new head
-// will be deleted and the new one set.
-func (hc *HeaderChain) SetHead(head uint64, updateFn UpdateHeadBlocksCallback, delFn DeleteBlockContentCallback) {
-	var (
-		parentHash common.Hash
-		batch      = hc.chainDb.NewBatch()
-		origin     = true
-	)
-	for hdr := hc.CurrentHeader(); hdr != nil && hdr.Number[types.QuaiNetworkContext].Uint64() > head; hdr = hc.CurrentHeader() {
-		num := hdr.Number[types.QuaiNetworkContext].Uint64()
-
-		// Rewind block chain to new head.
-		parent := hc.GetHeader(hdr.ParentHash[types.QuaiNetworkContext], num-1)
-		if parent == nil {
-			parent = hc.genesisHeader
-		}
-		parentHash = parent.Hash()
-
-		// Notably, since geth has the possibility for setting the head to a low
-		// height which is even lower than ancient head.
-		// In order to ensure that the head is always no higher than the data in
-		// the database (ancient store or active store), we need to update head
-		// first then remove the relative data from the database.
-		//
-		// Update head first(head fast block, head full block) before deleting the data.
-		markerBatch := hc.chainDb.NewBatch()
-		if updateFn != nil {
-			newHead, force := updateFn(markerBatch, parent)
-			if force && newHead < head {
-				log.Warn("Force rewinding till ancient limit", "head", newHead)
-				head = newHead
-			}
-		}
-		// Update head header then.
-		rawdb.WriteHeadHeaderHash(markerBatch, parentHash)
-		if err := markerBatch.Write(); err != nil {
-			log.Crit("Failed to update chain markers", "error", err)
-		}
-		hc.currentHeader.Store(parent)
-		hc.currentHeaderHash = parentHash
-		headHeaderGauge.Update(parent.Number[types.QuaiNetworkContext].Int64())
-
-		// If this is the first iteration, wipe any leftover data upwards too so
-		// we don't end up with dangling daps in the database
-		var nums []uint64
-		if origin {
-			for n := num + 1; len(rawdb.ReadAllHashes(hc.chainDb, n)) > 0; n++ {
-				nums = append([]uint64{n}, nums...) // suboptimal, but we don't really expect this path
-			}
-			origin = false
-		}
-		nums = append(nums, num)
-
-		// Remove the related data from the database on all sidechains
-		for _, num := range nums {
-			// Gather all the side fork hashes
-			hashes := rawdb.ReadAllHashes(hc.chainDb, num)
```

Comment:
Using your reasoning, the new design is needed since we cannot create sidechains in a meaningful way. It seems as if this ReadAllHashes can construct those sidechains, no?